ompi/mca/osc/ucx/osc_ucx_active_target.c

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2017. ALL RIGHTS RESERVED.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+
+#include "ompi/mca/osc/osc.h"
+#include "ompi/mca/osc/base/base.h"
+#include "ompi/mca/osc/base/osc_base_obj_convert.h"
+
+#include "osc_ucx.h"
+
+
+int ompi_osc_ucx_fence(int assert, struct ompi_win_t *win) {
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t*) win->w_osc_module;
+    int ret = OMPI_SUCCESS;
+    ucs_status_t status;
+
+    if (module->epoch_type.access != NONE_EPOCH &&
+        module->epoch_type.access != FENCE_EPOCH) {
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    if (assert & MPI_MODE_NOSUCCEED) {
+        module->epoch_type.access = NONE_EPOCH;
+    } else {
+        module->epoch_type.access = FENCE_EPOCH;
```

Comment:
I agree that fence calls must be used in pairs. however, the problem here is that the MPI library does not know if the current fence is used to close the epoch or open the epoch, if there is no assert.

For example,

rank0                      rank 1
fence(0, win)          fence(0, win)
put/get                   put/get
fence(0, win)          fence(0, win)
put/get                   put/get
fence(0, win)           fence(0, win)

In this example, from the perspective of MPI library, it cannot decide if the second pair of fence is used to close the epoch or open a new epoch. So, if there is no assert at all, the epoch state should always be FENCE_EPOCH until MPI library meets another sync call (PSCW, LOCK....).